File: WindowsApp1/DataEntry.vb

Imports System.Text.RegularExpressions
''' <summary>
''' PLC Database
''' An instance of this class stores all data of a RSS file. One database is binded to one file. To load a new file,
''' a new instance needs to be created.
''' </summary>
Public Class DataEntry
    Private addr As String
    Private name As String
    Private desp As String
    Private des As List(Of String)
    Private src As List(Of String)
    Private ext As String
    Private logic As Node
    Dim addr_format As New Regex("^(?:([A-Z]{1,3})(\d{1,3}):(\d{1,3})|(?:(I|O|S|U):(\d{1,3}(?:\.\d{1,3})*)))(?:\/(\d{1,2}))*(.*)$")
    Public Sub New(address As String, name As String, description As String)
        Dim match As Match = addr_format.Match(address)
        If Not match.Success Then
            Throw New ArgumentException("Invalid address format: " & address)
        End If
        addr = address
        Me.name = name
        desp = description
        des = New List(Of String)
        src = New List(Of String)
        ext = ""
    End Sub

    Public Sub New(address As String)
        Dim match As Match = addr_format.Match(address)
        If Not match.Success Then
            Throw New ArgumentException("Invalid address format: " & address)
        End If
        addr = address
        desp = ""
        name = ""
        des = New List(Of String)
        src = New List(Of String)
        ext = ""
    End Sub

    Public Sub CopyNameAndDesp(other As DataEntry)
        name = other.TagName
        desp = other.Description
    End Sub
    Public Property Address As String
        Get
            Return addr
        End Get
        Set(value As String)
            addr = value
        End Set
    End Property

    Public Property TagName As String
        Get
            Return name
        End Get
        Set(value As String)
            If name <> value Then
            End If
            name = value
        End Set
    End Property

    Public Property Description As String
        Get
            Return desp
        End Get
        Set(value As String)
            If desp <> value Then
            End If
            desp = value
        End Set
    End Property

    Public Property MappingTo As List(Of String)
        Get
            Return des
        End Get
        Set(value As List(Of String))
            des = value
        End Set
    End Property

    Public Property MappingLogic As Node
        Get
            Return logic
        End Get
        Set(value As Node)
            logic = value
        End Set
    End Property
    Public Sub AddMappingTo(addr As String)
        If Not MappingTo.Contains(addr) Then
            des.Add(addr)
        End If
    End Sub
    Public Sub AddMappedTo(addr As String)
        If Not MappedTo.Contains(addr) Then
            src.Add(addr)
        End If
    End Sub
    Public Property MappedTo As List(Of String)
        Get
            Return src
        End Get
        Set(value As List(Of String))
            src = value
        End Set
    End Property
    Public Property Extension As String
        Get
            Return ext
        End Get
        Set(value As String)
            ext = value
        End Set
    End Property
End Class

File: WindowsApp1/DataEntry.cs
using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;

namespace WindowsApp1
{
    /// <summary>
/// PLC Database
/// An instance of this class stores all data of a RSS file. One database is binded to one file. To load a new file,
/// a new instance needs to be created.
/// </summary>
    public class DataEntry
    {
        private string addr;
        private string name;
        private string desp;
        private List<string> des;
        private List<string> src;
        private string ext;
        private Node logic;
        private Regex addr_format = new Regex(@"^(?:([A-Z]{1,3})(\d{1,3}):(\d{1,3})|(?:(I|O|S|U):(\d{1,3}(?:\.\d{1,3})*)))(?:\/(\d{1,2}))*(.*)$");
        public DataEntry(string address, string name, string description)
        {
            var match = addr_format.Match(address);
            if (!match.Success)
            {
                throw new ArgumentException("Invalid address format: " + address);
            }
            addr = address;
            this.name = name;
            desp = description;
            des = new List<string>();
            src = new List<string>();
            ext = "";
        }

        public DataEntry(string address)
        {
            var match = addr_format.Match(address);
            if (!match.Success)
            {
                throw new ArgumentException("Invalid address format: " + address);
            }
            addr = address;
            desp = "";
            name = "";
            des = new List<string>();
            src = new List<string>();
            ext = "";
        }

        public void CopyNameAndDesp(DataEntry other)
        {
            name = other.TagName;
            desp = other.Description;
        }
        public string Address
        {
            get
            {
                return addr;
            }
            set
            {
                addr = value;
            }
        }

        public string TagName
        {
            get
            {
                return name;
            }
            set
            {
                if ((name ?? "") != (value ?? ""))
                {
                }
                name = value;
            }
        }

        public string Description
        {
            get
            {
                return desp;
            }
            set
            {
                if ((desp ?? "") != (value ?? ""))
                {
                }
                desp = value;
            }
        }

        public List<string> MappingTo
        {
            get
            {
                return des;
            }
            set
            {
                des = value;
            }
        }

        public Node MappingLogic
        {
            get
            {
                return logic;
            }
            set
            {
                logic = value;
            }
        }
        public void AddMappingTo(string addr)
        {
            if (!MappingTo.Contains(addr))
            {
                des.Add(addr);
            }
        }
        public void AddMappedTo(string addr)
        {
            if (!MappedTo.Contains(addr))
            {
                src.Add(addr);
            }
        }
        public List<string> MappedTo
        {
            get
            {
                return src;
            }
            set
            {
                src = value;
            }
        }
        public string Extension
        {
            get
            {
                return ext;
            }
            set
            {
                ext = value;
            }
        }
    }
}